### PR TITLE
Do not allow tables on different replicas have different aggregate functions in SimpleAggregateFunction column

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Do not allow tables on different replicas have different aggregate functions in SimpleAggregateFunction columns

Currently users can do this:

```sql
CREATE TABLE t
(
    `id` UInt64,
    `val` SimpleAggregateFunction(max, Nullable(String))
)
ENGINE = ReplicatedAggregatingMergeTree('/clickhouse/tables/{database}/t', 'r1')
ORDER BY id

Query id: 5ac61e5c-a72e-4c56-8b87-6ae543190a1a

Ok.

0 rows in set. Elapsed: 0.016 sec.
```

then on another replica

```sql
CREATE TABLE t
(
    `id` UInt64,
    `val` SimpleAggregateFunction(anyLast, Nullable(String))
)
ENGINE = ReplicatedAggregatingMergeTree('/clickhouse/tables/{database}/t', 'r2')
ORDER BY id


Query id: 3e069b14-0a3e-4adf-bc05-132ed576773e

Ok.

0 rows in set. Elapsed: 0.016 sec.
```

And this is a disaster if we don't find the issue fast enough.

tag @tavplubix 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/ColumnsDescription.cpp
Comment (alexey-milovidov) [inline src/Storages/ColumnsDescription.cpp:62]: Missing comments.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 57b160b6b9c7cdcf07efd70f947748b6a813f2f7 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56724/57b160b6b9c7cdcf07efd70f947748b6a813f2f7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6860414443">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56724/57b160b6b9c7cdcf07efd70f947748b6a813f2f7/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/ColumnsDescription.cpp
Comment (alexey-milovidov) [inline src/Storages/ColumnsDescription.cpp:61]: This makes the code very difficult to understand.

File: src/Storages/ColumnsDescription.cpp
Comment (alexey-milovidov) [reply src/Storages/ColumnsDescription.cpp:62]: The code is wrong because it applies not only to Replicated tables,
but we allow aggregate functions with compatible states for Distributed, Merge, and Buffer tables.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:1235]: Looks ok to me, but I also need to ask @alesapin or @tavplubix

**Comment by canhld94** [comment]
And honestly it was a bug to me. Allow people to do bad thing can be considered a bug.
@tavplubix @alesapin pls help to review when you havr time [THANKS]

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:1234]: > data types like SimpleAggregateFunction may have different aggregate function in 1st argument but still compatible if 2nd argument is same

Why? This looks like the root cause, and probably we should fix SimpleAggregateFunction types comparison instead

Comment (canhld94) [reply src/Storages/StorageReplicatedMergeTree.cpp:1234]: I don't think we should do that.  Refer to Alexey comment above when I changed comparison logic for `ColumnDescription`

> The code is wrong because it applies not only to Replicated tables,
> but we allow aggregate functions with compatible states for Distributed, Merge, and Buffer tables.

So basically as far as I understand, except for replicated table, we want `SimpleAggregateFunction` to be compatible (as long as 2nd arguments are same) in almost all other cases.

Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:1234]: > The code is wrong because it applies not only to Replicated tables

Well, I agree that the code was wrong but definitely not because of this :) 
The right way was to fix the `equals` method for `SimpleAggregateFunction`, not to compare type names as strings

> but we allow aggregate functions with compatible states for Distributed, Merge, and Buffer tables.

Why would we allow that? Or why would anyone use that? Looks like a way to shoot your leg.

> So basically as far as I understand, except for replicated table, we want SimpleAggregateFunction to be compatible (when 2nd arguments are same) in almost all other cases.

I don't know what Alexey wants :) 

> And this is a disaster if we don't find the issue fast enough.

What is the disaster exactly?

Comment (canhld94) [reply src/Storages/StorageReplicatedMergeTree.cpp:1234]: > The right way was to fix the equals method for SimpleAggregateFunction, not to compare type names as strings

From implementation POV, `SimpleAggegateFunction` is not really a `IDataType` but rather an ordinary `IDataType` with some custom name and serialisation:

```cpp
  /// This is mutable to allow setting custom name and serialization on `const IDataType` post construction.
  mutable DataTypeCustomNamePtr custom_name;
  mutable SerializationPtr custom_serialization;
```

So I guess you want to compare `custom_name` if they're non-null? The scope will be broader than just to fix the issue in this PR. Waiting for final decision whether we should do this or not.

> Why would we allow that? Or why would anyone use that? Looks like a way to shoot your leg.

May be @alexey-milovidov can answer this question :)

> What is the disaster exactly?

Data inconsistency - some aggregated with `max`, and some aggregated with `anyLast`, for example.
Replicated merges have logic to check checksum and download to make part to be byte-identical, but we can't control which replica will download from which replica. In reality, we observe that both replicas download from the other to fix checksum mismatch. So in the end all replicas have mixed data.

**Comment by alexey-milovidov** [comment]
1. `SimpleAggregaFunction` data types should be considered compatible if the return type of the aggregate function (the storage representation) is the same.
2. Replicas should check not just for compatible data types but more strictly - for them to be fully identical.

For example, for the usual AggregateFunction, `quantile(0.5)` and `quantile(0.9)` are compatible, and they can be used interchangingly in Buffer, Merge, and Distributed tables. But in ReplicatedMergeTree, they have to be identical.

**Comment by canhld94** [comment]
Just for FYI, we found that users can create tables in different replicas with different merge tree engine parameters. Should be fixed in another PR.

**Comment by Algunenano** [comment]
> Just FYI, we found that users can create tables in different replicas with different merge tree engine parameters. Should be fixed in another PR.

What makes you think that's a bug? There are good reasons why somebody might use different settings in different replicas, and that's ok so long you know what you are doing.

**Comment by canhld94** [comment]
> What makes you think that's a bug? There are good reasons why somebody might use different settings in different replicas, and that's ok so long you know what you are doing.

Yes, different settings is normal. I mean the merge tree engine parameter.
For example, today I found one of my backend developers create two replicas using ReplicatedReplacingMergeRree, one with version column, one without. And CH didn't complain about that. And the result is merges consistently failing with CHECKSUM_DOESNT_MATCH, resulting in high network traffic between two replicas to sync.
I see that CH does check consistency for sign columns, so it should also check for other columns.
@Algunenano @alexey-milovidov does this make sense to you?

**Comment by alexey-milovidov** [comment]
Yes, the parameters related to merges should be checked for identity.

**Comment by canhld94** [comment]
> Yes, the parameters related to merges should be checked for identity.

Create another PR for this #56833

> 1. SimpleAggregaFunction data types should be considered compatible if the return type of the aggregate function (the storage representation) is the same.
> 2. Replicas should check not just for compatible data types but more strictly - for them to be fully identical.

> For example, for the usual AggregateFunction, quantile(0.5) and quantile(0.9) are compatible, and they can be used interchangingly in Buffer, Merge, Materialized View, View, and Distributed tables. But in ReplicatedMergeTree, they have to be identical.

@tavplubix wdyt? In order to make type to be "identical", current approach in this PR is to compare type names. May be we add a method `isIdentical` to `IDataType`?

**Comment by tavplubix** [comment]
> In order to make type to be "identical", current approach in this PR is to compare type names. May be we add a method isIdentical to IDataType?

Yes, I think we should introduce a new method for this, or just add an ad-hoc function that compares `SimpleAggregaFunction` in a more strict way. And use this method/function when comparing structures for ReplicatedMergeTree specifically.

Why it's a bad idea to simply compare types as strings: https://github.com/ClickHouse/ClickHouse/issues/44477

**Comment by canhld94** [comment]
> Yes, I think we should introduce a new method for this, or just add an ad-hoc function that compares `SimpleAggregaFunction` in a more strict way. And use this method/function when comparing structures for ReplicatedMergeTree specifically.
> 
> Why it's a bad idea to simply compare types as strings: #44477

Updated implementation

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@canhld94, The bug that we've found is something as follows:
- All GIS data types look like they are not identical to the instance of the same data type:
`Type: MultiPolygon, other: MultiPolygon, identical?: 0, equals?: 1`

**Pull request merged**
